lib/Graph/Graph.cpp

```diff
@@ -2808,42 +2808,58 @@ CollectRpnProposalsNode *Function::createCollectRpnProposals(
   ShapeVector outDims;
-  outDims.insert(outDims.end(), dDims.begin(), dDims.begin() + batchDims);
+  outDims.insert(outDims.end(), dDims.begin(), dDims.begin() + axis);
   outDims.insert(outDims.end(), iDims.begin(), iDims.end());
-  outDims.insert(outDims.end(), dDims.begin() + batchDims + 1, dDims.end());
+  outDims.insert(outDims.end(), dDims.begin() + axis + 1, dDims.end());
   return addNode(new GatherNode(
       name, getParent()->uniqueTypeWithNewShape(data.getType(), outDims), data,
-      indices, batchDims));
+      indices, axis));
 }
 
 GatherNDNode *Function::createGatherND(llvm::StringRef name, NodeValue data,
-                                       NodeValue indices) {
-  auto dDims = data.dims();
-  auto iDims = indices.dims();
-  int64_t lastIndicesDimension = iDims[iDims.size() - 1];
-  assert(dDims.size() > 0 && "data/input rank must be >= 1.");
-  assert(iDims.size() > 0 && "indices rank must be >= 1.");
-  assert(iDims[iDims.size() - 1] <= dDims.size() &&
-         "last dimension of indices can be at most rank of data/input");
-
-  int64_t outRank = dDims.size() + iDims.size() - lastIndicesDimension - 1;
-  std::vector<dim_t> outDimsGatherND(outRank);
-  size_t i = 0;
-  for (; i < iDims.size() - 1; i++) {
-    outDimsGatherND[i] = iDims[i];
-  }
-  for (size_t j = lastIndicesDimension; j < dDims.size(); i++, j++) {
-    outDimsGatherND[i] = dDims[j];
+                                       NodeValue indices,
+                                       unsigned_t batchDims) {
+  auto dataDims = data.dims();
+  auto indicesDims = indices.dims();
+  size_t indicesDimLast = indicesDims.back();
+
+  // Validate the input dimensions.
+  assert(dataDims.size() >= 1 && "Input data rank must be >= 1.");
+  assert(indicesDims.size() >= 1 && "Input indices rank must be >= 1.");
+  for (size_t idx = 0; idx < batchDims; ++idx) {
+    assert(dataDims[idx] == indicesDims[idx] &&
+           "Batch dimensions of data and indices must be the same!");
+  }
+  assert(batchDims < std::min(dataDims.size(), indicesDims.size()) &&
+         "The number of batch dimensions must be smaller than both the input "
+         "data and indices rank!");
+  assert(indicesDimLast >= 1 &&
+         "Last dimension of indices must be at least 1!");
+  assert(indicesDimLast <= (dataDims.size() - batchDims) &&
+         "Last dimension of indices must be at most rank of data without batch "
+         "dimensions!");
+
+  // Compute the output dimensions.
+  size_t outRank =
+      dataDims.size() + indicesDims.size() - indicesDimLast - 1 - batchDims;
+  std::vector<dim_t> outDims(outRank);
+  size_t outIdx = 0;
+  for (size_t idx = 0; idx < batchDims; ++idx) {
+    outDims[outIdx++] = dataDims[idx];
+  }
+  for (size_t idx = batchDims; idx < indicesDims.size() - 1; ++idx) {
+    outDims[outIdx++] = indicesDims[idx];
+  }
+  for (size_t idx = batchDims + indicesDimLast; idx < dataDims.size(); ++idx) {
+    outDims[outIdx++] = dataDims[idx];
```

Comment:
@khabinov please help review this part wrt semantics of GatherND. Rest part lgtm.